core/src/main/scala/org/apache/spark/sql/delta/stats/DataSkippingStatsTracker.scala

```diff
@@ -0,0 +1,203 @@
+  private val extendedRow: GenericInternalRow = new GenericInternalRow(2)
+
+  // file path to corresponding stats encoded as json
+  protected val results = new collection.mutable.HashMap[String, String]
+
+  // called once per file, executes the getStats projection
+  override def closeFile(filePath: String): Unit = {
+    // We assume file names are unique
+    val fileName = new Path(filePath).getName
+
+    assert(!results.contains(fileName), s"Stats already recorded for file: $filePath")
+    // this is statsColExpr's output (json string)
+    val jsonStats = getStats(submittedFiles(filePath)).getString(0)
+    results += ((fileName, jsonStats))
+    submittedFiles.remove(filePath)
+  }
+
+  override def newPartition(partitionValues: InternalRow): Unit = { }
+
+  override def newFile(newFilePath: String): Unit = {
+    submittedFiles.getOrElseUpdate(newFilePath, {
+      // `buffer` is a row that will start off by holding the initial values for the agg expressions
+      // (see the initializeStats: Projection), will then be updated in place every time a new row
+      // is processed (see updateStats: Projection), and will finally serve as an input for
+      // computing the per-file result of statsColExpr (see getStats: Projection)
+      val buffer = new SpecificInternalRow(aggBufferAttrs.map(_.dataType))
+      buffer
+    })
+  }
+
+  override def newRow(filePath: String, currentRow: InternalRow): Unit = {
+    val aggBuffer = submittedFiles(filePath)
+    extendedRow.update(0, aggBuffer)
+    extendedRow.update(1, currentRow)
+    updateStats.target(aggBuffer).apply(extendedRow)
+  }
+
+  override def getFinalStats(taskCommitTime: Long): DeltaFileStatistics = {
+    submittedFiles.keys.foreach(closeFile)
+    submittedFiles.clear()
+    DeltaFileStatistics(results.toMap)
+  }
+
+  /**
+   * Flattens the struct nested columns and for each column returns its data type and an unique
+   * custom name created by appending the inner field column name to the given `customExprId`
+   * (i.e. an accumulator for so far created custom name, which initially is equal
+   * to the column `exprId`).
+   */
+  def flattenStructCols(field: StructType, customExprId: String): Array[(String, DataType)] = {
+    field.fields.flatMap { f =>
+      f.dataType match {
+        case structField: StructType =>
+          flattenStructCols(structField, customExprId + "." + f.name)
+        case _ =>
+          Array((customExprId + "." + f.name, f.dataType))
+      }
+    }
+  }
+}
+
+/**
+ * Serializable factory class that holds together all required parameters for being able to
+ * instantiate a [[DeltaStatisticsTracker]] on an executor.
+ *
+ * @param hadoopConf The Hadoop configuration object to use on an executor.
+ * @param path Root Reservoir path
+ * @param dataCols Resolved data (i.e. non-partitionBy) columns of the dataframe to be written.
+ */
+class DeltaJobStatisticsTracker(
```

Comment:
Its not obvious from the names how DeltaStatisticsTracker is different from DeltaJobStatisticsTracker. Maybe rename DeltaStatisticsTracker to DeltaTaskStatisticsTracker